tariff/octopus.go

```diff
@@ -0,0 +1,101 @@
+package tariff
+
+import (
+	"errors"
+	"github.com/evcc-io/evcc/tariff/octopus"
+	"sync"
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/request"
+)
+
+type Octopus struct {
+	mux     sync.Mutex
+	log     *util.Logger
+	uri     string
+	region  string
+	data    api.Rates
+	updated time.Time
+}
+
+var _ api.Tariff = (*Octopus)(nil)
+
+func init() {
+	registry.Add("octopusenergy", NewOctopusFromConfig)
+}
+
+func NewOctopusFromConfig(other map[string]interface{}) (api.Tariff, error) {
+	cc := struct {
+		Region string
+		Tariff string
+	}{}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	if cc.Region == "" {
+		return nil, errors.New("missing region")
+	}
+	if cc.Tariff == "" {
+		return nil, errors.New("missing tariff code")
+	}
+
+	t := &Octopus{
+		log:    util.NewLogger("octopus"),
+		uri:    octopus.ConstructRatesAPI(cc.Tariff, cc.Region),
+		region: cc.Tariff,
+	}
+
+	done := make(chan error)
+	go t.run(done)
+	err := <-done
+
+	return t, err
+}
+
+func (t *Octopus) run(done chan error) {
+	var once sync.Once
+	client := request.NewHelper(t.log)
+
+	for ; true; <-time.NewTicker(time.Hour).C {
+		var res octopus.UnitRates
+		if err := client.GetJSON(t.uri, &res); err != nil {
+			t.log.ERROR.Println(err)
```

Comment:
Think this needs an

			once.Do(func() { done <- err })

Seems I forgot this everywhere

Reply:
In theory, wouldn't we only return an error on the pipe if we wanted `NewOctopusFromConfig` to return an error? Would that be expected behaviour for something like a JSON parse failure? We already log the failure to the object logger

Reply:
We are reading a single result from the channel- either error or nil. This value is also only written once. Yes, if the initial query fails we return and log the error both which is not necessary.
I still like the pattern since we can return a single (initial) result from the goroutine. If there's a better alternative, happy to adopt that.

Reply:
This makes more sense with coffee in me, LGTM (and fixed in your commit)